[WPE] Add wp_presentation_feedback handling and forward data into sysprof

nikolaszimmermann opened the pull request:
#### 31a6f52402f8ffbb840b3d1d1d81dc0194c69923
<pre>
[WPE] Add wp_presentation_feedback handling and forward data into sysprof
<a href="https://bugs.webkit.org/show_bug.cgi?id=293343">https://bugs.webkit.org/show_bug.cgi?id=293343</a>

Reviewed by Carlos Garcia Campos.

Implement Wayland wp_presentation_feedback protocol handling and
intercept sync_output / presented / discarded events. Those events can
be used to verify how long it took from frame generation to presentation
on screen, and get post-display statistics on latency / frame drops etc.

We only collect those information when sysprof tracing is active to
avoid any overhead if we don&apos;t need it.

* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandConstructed):
(wpeDisplayWaylandDispose):
(wpeDisplayWaylandGetPresentation):
(wpe_display_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandDispose):
(PresentationFeedbackStatistics::~PresentationFeedbackStatistics):
(PresentationFeedbackStatistics::beginFrame):
(PresentationFeedbackStatistics::presentedFrame):
(PresentationFeedbackStatistics::discardedFrame):
(PresentationFeedbackStatistics::calculateAverageLatency const):
(wpeViewWaylandRenderBuffer): Deleted.

Canonical link: <a href="https://commits.webkit.org/297679@main">https://commits.webkit.org/297679@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/edaf177a3cb53f48d930c268d6c48f9e46541dd3

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/112537 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32269 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/22747 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/118735 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/62991 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/32921 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/40832 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/85650 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/36273 "") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/115484 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26276 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101249 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65952 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25575 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/62495 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/95667 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19455 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/121957 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/39611 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29506 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94517 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/39993 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97481 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94257 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24059 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39374 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17179 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/35699 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/39499 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/44987 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39136 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/42471 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/40877 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/45bec95f0580a25948f30f09865703ab5abb6458. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45684

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1d132838f4988268732f2c19611cb570bf1ecda0. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/45684

GeorgesStavracas reviewed (approved):
The Sysprof bits LGTM. The counter is being used correctly, despite the misreported data arriving on Sysprof. This is probably the old cross-process unique id clash issue that I've been trying to fix for some time now.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
```diff
@@ -129,6 +130,19 @@ add_custom_command(
     COMMAND ${WAYLAND_SCANNER} client-header ${WAYLAND_PROTOCOLS_DATADIR}/unstable/pointer-constraints/pointer-constraints-unstable-v1.xml ${WPEPlatform_DERIVED_SOURCES_DIR}/wpe/wayland/pointer-constraints-unstable-v1-client-protocol.h
     VERBATIM)
 
+add_custom_command(
+    OUTPUT ${WPEPlatform_DERIVED_SOURCES_DIR}/wpe/wayland/presentation-time-protocol.c
+    MAIN_DEPENDENCY ${WAYLAND_PROTOCOLS_DATADIR}/stable/presentation-time/presentation-time.xml
+    DEPENDS ${WPEPlatform_DERIVED_SOURCES_DIR}/wpe/wayland/presentation-time-client-protocol.h
+    COMMAND ${WAYLAND_SCANNER} private-code ${WAYLAND_PROTOCOLS_DATADIR}/stable/presentation-time/presentation-time.xml ${WPEPlatform_DERIVED_SOURCES_DIR}/wpe/wayland/presentation-time-protocol.c
+    VERBATIM)
+
+add_custom_command(
+    OUTPUT ${WPEPlatform_DERIVED_SOURCES_DIR}/wpe/wayland/presentation-time-client-protocol.h
+    MAIN_DEPENDENCY ${WAYLAND_PROTOCOLS_DATADIR}/stable/presentation-time/presentation-time.xml
+    COMMAND ${WAYLAND_SCANNER} client-header ${WAYLAND_PROTOCOLS_DATADIR}/stable/presentation-time/presentation-time.xml ${WPEPlatform_DERIVED_SOURCES_DIR}/wpe/wayland/presentation-time-client-protocol.h
+    VERBATIM)
+
```
Can we do this only when building with sysprof enabled?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
```diff
@@ -86,6 +88,7 @@ struct _WPEDisplayWaylandPrivate {
     struct zwp_text_input_v3* textInputV3;
     struct zwp_pointer_constraints_v1* pointerConstraints;
     struct zwp_relative_pointer_manager_v1* relativePointerManager;
+    struct wp_presentation* presentation;
```
Same here and all other places where you didn't add the ifdef.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
```diff
@@ -178,6 +181,14 @@ static GRefPtr<GSource> wpeDisplayWaylandCreateEventSource(WPEDisplayWayland* di
+static void wpeDisplayWaylandConstructed(GObject* object)
+{
+    G_OBJECT_CLASS(wpe_display_wayland_parent_class)->constructed(object);
+#if USE(SYSPROF_CAPTURE)
+    SysprofAnnotator::createIfNeeded("WPE/Wayland Platform"_s);
+#endif
```
Don't we have one singleton per process? This is the UI process, so I guess "WebKit (UI)" already exists and this does nothing.

carlosgcampos replied:
I'm sorry, but I still don't understand this. We expect to have one annotator per process, and process name is what createIfNeeded expects. We are already calling  `SysprofAnnotator::createIfNeeded` from `webkitInitialize()` passing "WebKit (UI)". WPEDisplay can only be created in the UI process, but it can happen that is created before the webkit web view or web context. But in that case we would be creating the annotator before webkit, so webkit marks will appear as in the process "WPE/Wayland Platform".

nikolaszimmermann replied:
I didn't investigate, I only checked that I need the line to get marks from WPEDisplayWayland. Will check tomorrow, maybe @GeorgesStavracas has an idea as well.

nikolaszimmermann replied:
As discussed, the symbol is duplicated:

```
📦 nzimmermann@wkdev:/host/home/nzimmermann/Software/GitRepositories/WebKit/WebKitBuild/WPE/Release/$ objdump -x lib/libWPEPlatform-2.0.so.0.0.0 | c++filt | grep -i createIfNeed
00000000002c6990 l     F .text	00000000000000b5              .hidden std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::{lambda()#1}>(std::once_flag&, WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::{lambda()#1}&&)::{lambda()#1}>(WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::{lambda()#1}&)::{lambda()#1}::__invoke()
0000000000333d48 l     O .bss	0000000000000004              .hidden WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::onceFlag
0000000000333d10 l     O .bss	0000000000000038              .hidden WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::instance
```

```
📦 nzimmermann@wkdev:/host/home/nzimmermann/Software/GitRepositories/WebKit/WebKitBuild/WPE/Release/$ objdump -x lib/libWPEWebKit-2.0.so.1.6.2 | c++filt | grep -i createIfNeed
0000000004bc4f90 l     F .text	00000000000000b5              .hidden std::once_flag::_Prepare_execution::_Prepare_execution<std::call_once<WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::{lambda()#1}>(std::once_flag&, WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::{lambda()#1}&&)::{lambda()#1}>(WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::{lambda()#1}&)::{lambda()#1}::__invoke()
000000000b753698 l     O .bss	0000000000000004              .hidden WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::onceFlag
000000000b753660 l     O .bss	0000000000000038              .hidden WTF::SysprofAnnotator::createIfNeeded(WTF::ASCIILiteral)::instance
```

So there is no shared SysprofAnnotator between anything in UIProces and stuff in libWPEPlatform*

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+static const struct wp_presentation_feedback_listener presentationFeedbackListener = {
+    // sync_output
+    [](void*, struct wp_presentation_feedback*, struct wl_output*) {
+        // Quoting https://wayland.app/protocols/presentation-time#wp_presentation_feedback:event:presented:
+        // As presentation can be synchronized to only one output at a time, this event tells which output it was.
+        // This event is only sent prior to the presented event.
```
Don't we need to check the output here is the one where the view is?

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+        // This event is only sent prior to the presented event.
+    },
+    // presented
+    [](void* data, struct wp_presentation_feedback* feedback, uint32_t tv_sec_hi, uint32_t tv_sec_lo, uint32_t tv_nsec, uint32_t refresh_nsec, uint32_t seq_hi, uint32_t seq_lo, uint32_t flags) {
```
```suggestion
    [](void* data, struct wp_presentation_feedback* feedback, uint32_t tvSecHi, uint32_t tvSecLo, uint32_t tvNsec, uint32_t refreshNsec, uint32_t seqHi, uint32_t seqLo, uint32_t flags) {
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+    },
+    // presented
+    [](void* data, struct wp_presentation_feedback* feedback, uint32_t tv_sec_hi, uint32_t tv_sec_lo, uint32_t tv_nsec, uint32_t refresh_nsec, uint32_t seq_hi, uint32_t seq_lo, uint32_t flags) {
+        auto* view = WPE_VIEW(data);
+        auto* priv = WPE_VIEW_WAYLAND(view)->priv;
```
```suggestion
        auto* priv = WPE_VIEW_WAYLAND(data)->priv;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+    },
+    // discarded
+    [](void* data, struct wp_presentation_feedback* feedback) {
+        auto* view = WPE_VIEW(data);
+        auto* priv = WPE_VIEW_WAYLAND(view)->priv;
```
```suggestion
        auto* priv = WPE_VIEW_WAYLAND(data)->priv;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+void PresentationFeedbackStatistics::beginFrame(struct wp_presentation* presentation, WPEView* view)
+{
+    auto* wlSurface = wpe_view_wayland_get_wl_surface(WPE_VIEW_WAYLAND(view));
+
```
```suggestion
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+void PresentationFeedbackStatistics::presentedFrame(struct wp_presentation_feedback* feedback, Seconds presentationTime, Seconds refreshInterval, uint64_t sequence, uint32_t flags)
+{
+    std::optional<Seconds> frameStartTime;
+    for (size_t i = 0; i < m_pendingFeedbacks.size(); ++i) {
+        auto [presentationFeedback, frameStartTimeForFeedback] = m_pendingFeedbacks[i];
+        if (presentationFeedback == feedback) {
+            frameStartTime = frameStartTimeForFeedback.secondsSinceEpoch();
+            wp_presentation_feedback_destroy(feedback);
+            m_pendingFeedbacks.remove(i);
+            break;
+        }
+    }
```
I think you can use removeFirstMatching() capturing frameStartTime in the lambda and calling wp_presentation_feedback_destroy() from the lambda too.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+        }
+    }
+
+    RELEASE_ASSERT(frameStartTime.has_value());
```
removeFirstMatching returns bool, so we can assert it always returns true and then we don't need to use optional for frameStartTime

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -330,6 +363,131 @@ const struct wl_callback_listener frameListener = {
+
+void PresentationFeedbackStatistics::discardedFrame(struct wp_presentation_feedback* feedback)
+{
+    for (size_t i = 0; i < m_pendingFeedbacks.size(); ++i) {
+        auto [presentationFeedback, frameStartTimeForFeedback] = m_pendingFeedbacks[i];
+        if (presentationFeedback == feedback) {
+            wp_presentation_feedback_destroy(feedback);
+            m_pendingFeedbacks.remove(i);
+            break;
+        }
+    }
```
removeFirstMatching() could be used here too.

nikolaszimmermann committed:
[WPE] Add wp_presentation_feedback handling and forward data into sysprof
https://bugs.webkit.org/show_bug.cgi?id=293343

Reviewed by Carlos Garcia Campos.

Implement Wayland wp_presentation_feedback protocol handling and
intercept sync_output / presented / discarded events. Those events can
be used to verify how long it took from frame generation to presentation
on screen, and get post-display statistics on latency / frame drops etc.

We only collect those information when sysprof tracing is active to
avoid any overhead if we don't need it.

* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandConstructed):
(wpeDisplayWaylandDispose):
(wpeDisplayWaylandGetPresentation):
(wpe_display_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(wpeViewWaylandDispose):
(PresentationFeedbackStatistics::~PresentationFeedbackStatistics):
(PresentationFeedbackStatistics::beginFrame):
(PresentationFeedbackStatistics::presentedFrame):
(PresentationFeedbackStatistics::discardedFrame):
(PresentationFeedbackStatistics::calculateAverageLatency const):
(wpeViewWaylandRenderBuffer): Deleted.

Canonical link: https://commits.webkit.org/297679@main

webkit-commit-queue commented:
Committed 297679@main (31a6f52402f8): <https://commits.webkit.org/297679@main>

Reviewed commits have been landed. Closing PR #45684 and removing active labels.
